src/video/rpi_cam.cc

```diff
@@ -0,0 +1,115 @@
+// BoB robotics includes
+#include "common/macros.h"
+#include "video/rpi_cam.h"
+
+// Standard C includes
+#include <cstring>
+
+// Standard C++ includes
+#include <stdexcept>
+
+// Extra POSIX includes
+#ifdef unix
+#include <fcntl.h>
+#endif
+
+namespace BoBRobotics {
+namespace Video {
+
+RPiCamera::RPiCamera(uint16_t port)
+  : m_Socket(INVALID_SOCKET)
+{
+    m_Port = port;
+
+    // set up the networking code needed to receive images
+    setupSockets();
+}
+
+std::string
+RPiCamera::getCameraName() const
+{
+    return "Raspberry Pi Camera";
+}
+
+cv::Size
+RPiCamera::getOutputSize() const
+{
+    // this is fixed
+    return cv::Size(152, 72);
+}
+
+bool
+RPiCamera::readFrame(cv::Mat &outFrame)
+{
+    unsigned char buffer[72 * 19];
+
+    // Check we're connected
+    BOB_ASSERT(m_Socket != INVALID_SOCKET);
+
+    // Make sure output frame is the right size and type
+    outFrame.create(72, 152, CV_32FC1);
+
+    // get the most recent UDP frame (grayscale for now)
+    while (recv(m_Socket, buffer, 72 * 19, 0) > 0) {
+        // fill in the outFrame
+        //std::cout << (int) buffer[0] << std::endl;
+        int j = 0;
+        for (int i = 0; i < 72 * 19 - 1; ++i) {
+            outFrame.at<float>(i % 72, buffer[0] + floor(i / 72)) = float(buffer[j]) / 255.0f;
```

Comment:
What's happening here!? 

1. I guess each frame is made up of 8 packets but I can't really figure out what this loop does as it doesn't seem to increment anywhere. 
2. All other cameras produce 8UC format images - this one should too
3. ``readFrame`` produces CV_8UC3 for all other cameras so, for this one, it should call ``readGreyscaleFrame`` and duplicate the greyscale across all channels

Reply:
Good point.

Reply:
We should also probably document the type of ``cv::Mat`` you get from these functions

Reply:
True :smile: